Source/DafnyCore/AST/PreprocessorHelper.cs

```diff
@@ -0,0 +1,196 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Diagnostics.Contracts;
+using System.IO;
+using System.Text;
+
+namespace Microsoft.Dafny; 
+
+// Dafny files can use preprocessing directives, e.g.
+/// ```
+/// method TEst() {
+/// #if UNDEFINEDVARIABLE
+///   assert false;
+/// #else
+///   assert true;
+/// #endif
+/// }
+/// ```
+/// is transformed into
+/// ```
+/// method TEst() {
+/// 
+/// 
+/// 
+///   assert true;
+/// 
+/// }
+/// ```
+/// 
+/// However, at this moment, there is no way to tell Dafny which preprocessing variables
+/// are defined, so only all other blocks than "else" are replaced by empty newlines
+/// Note that, because this process replaces newlines, this version of PreprocessorHelper
+/// recover existing newlines to ensure that, if there is no pre-processing directives,
+/// the program string is exactly the same as the original one.
+public static class PreprocessorHelper {
+  struct ReadState {
+    public bool hasSeenElse;
+    public bool mayStillIncludeAnotherAlternative;
+
+    public ReadState(bool hasSeenElse, bool mayStillIncludeAnotherAlternative) {
+      this.hasSeenElse = hasSeenElse;
+      this.mayStillIncludeAnotherAlternative = mayStillIncludeAnotherAlternative;
+    }
+  }
+
+  // "arg" is assumed to be trimmed
+  private static bool IfdefConditionSaysToInclude(string arg, List<string> /*!*/ defines) {
+    Contract.Requires(arg != null);
+    Contract.Requires(cce.NonNullElements(defines));
+    bool sense = true;
+    while (arg.StartsWith("!")) {
+      sense = !sense;
+      arg = arg.Substring(1).TrimStart();
+    }
+
+    return defines.Contains(arg) == sense;
+  }
+
+  public static string Fill(Stream stream, List<string> /*!*/ defines, string newline) {
+    Contract.Requires(stream != null);
+    Contract.Requires(cce.NonNullElements(defines));
+    Contract.Ensures(Contract.Result<string>() != null);
+    StreamReader /*!*/
+      reader = new StreamReader(stream);
+    var o = stream.CanSeek;
+    return Fill(reader, defines, newline);
+  }
+
+  public static string Fill(TextReader reader, List<string> /*!*/ defines, string newline) {
```

Comment:
All this file was taken from Boogie's original preprocessor helper, but I'll continue the renaming anyway because I think we don't need to rely on our previous dependency.